src/librustc_borrowck/diagnostics.rs

```diff
@@ -314,6 +314,59 @@ let c = &i; // still ok!
 ```
 "##,
 
+E0504: r##"
+This error occurs when an attempt is made to move a borrowed variable into a
+closure. For example:
+
+```compile_fail
+use std::thread;
+
+struct FancyNum {
+    num: u8
+}
+
+fn main() {
+    let fancy_num = FancyNum { num: 5 };
+    let fancy_ref = &fancy_num;
+
+    thread::spawn(move || {
+        println!("child thread: {}", fancy_num.num);
+        // error: cannot move `fancy_num` into closure because it is borrowed
+    }).join().expect("thread failed to join");
+
+    println!("main thread: {}", fancy_ref.num);
+}
+```
+
+Here, `fancy_num` is borrowed by `fancy_ref` and so cannot be moved into
+the closure given to `thread::spawn`. There is no way to move a value into
+a closure while it is borrowed, as that would invalidate the borrow.
+
+If the value has to be borrowed, try limiting the lifetime of the borrow using
```

Comment:
I've been messing with it a bit, and I don't understand this one. What you did obviously works, but I'm not sure what `Arc` or `RefCell` contribute. If the user only needed a reference (not a moved value) inside the closure, they could use one directly [like this](https://play.rust-lang.org/?gist=e92204d83661cdba1f06c709c0f26e51&version=stable&backtrace=0).

Moreover, in the thread example, your code above doesn't work because a `RefCell` of a reference isn't `Sync` ([Example here (this is just your code with the `let x = ` swapped for a `thread::spawn`](https://play.rust-lang.org/?gist=2956057dc16c3a9ddb9826b42c1eebeb&version=stable&backtrace=0)).

Perhaps we should guide them towards considering a reference or a `RefCell` (if they need non-concurrent shared mutable access) rather than a moved value.